piksi_tools/console/settings_view.py

```diff
@@ -100,44 +131,53 @@ def __init__(self, name, section, value, ordering, settings):
         self.notes = settings.settings_yaml.get_field(section, name, 'Notes')
         self.default_value = settings.settings_yaml.get_field(
             section, name, 'default value')
+        readonly = settings.settings_yaml.get_field(section, name, 'readonly')
+        # get_field returns empty string if field missing, so I need this check to assign bool to traits bool
+        if readonly:
+            self.readonly = True
         self.revert_in_progress = False
-        self.confirmed_set = False
-
-    def revert_after_delay(self, delay, name, old, new, popop=True):
-        '''Revert setting to old value after delay in seconds has elapsed'''
-        time.sleep(delay)
-        if not self.confirmed_set:
-            self.revert_in_progress = True
-            self.value = old
-            invalid_setting_prompt = prompt.CallbackPrompt(
-                title="Settings Write Error",
-                actions=[prompt.close_button], )
-            invalid_setting_prompt.text = \
-                ("\n   Unable to confirm that {0} was set to {1}.\n"
-                 "   Ensure the range and formatting of the entry are correct.\n"
-                 "   Ensure that the new setting value did not interrupt console communication.").format(self.name, new)
-            invalid_setting_prompt.run()
-
-        self.confirmed_set = False
+        self.timed_revert_thread = None
+        self.confirmed_set = True
+
+    def revert_to_prior_value(self, name, old, new):
+        '''Revert setting to old value in the case we can't confirm new value'''
+        self.revert_in_progress = True
+        self.value = old
+        self.revert_in_progress = False
+        self.confirmed_set = True
+        invalid_setting_prompt = prompt.CallbackPrompt(
+            title="Settings Write Error",
+            actions=[prompt.close_button], )
+        invalid_setting_prompt.text = \
+            ("\n   Unable to confirm that {0} was set to {1}.\n"
+             "   Ensure the range and formatting of the entry are correct.\n"
+             "   Ensure that the new setting value did not interrupt console communication.").format(self.name, new)
+        invalid_setting_prompt.run()
 
     def _value_changed(self, name, old, new):
+        '''When a user changes a value, kick off a timed revert thread to revert it in GUI if no confirmation
+           that the change was successful is received. '''
         if (old != new and old is not Undefined and new is not Undefined):
             if type(self.value) == unicode:
                 self.value = self.value.encode('ascii', 'replace')
+            # Revert_in_progress is a guard to prevent this from running when the revert function changes the value
             if not self.revert_in_progress:
```

Comment:
does this variable need to be protected with a lock?

Reply:
Yes it does, otherwise when the settings value is reverted, it will kick off another thread to confirm the value.  This way I limit the recursion depth to 1.